### PR TITLE
feat: implement DataBlockBuilderImpl for variable width data block

### PR DESCRIPTION
**Pull request by broccoliSpicy**
This PR implements `DataBlockBuilderImpl` trait for Variable Width Data, so we can enable read Variable Width data in Lance 2.1

### CODE REVIEW

File: rust/lance-encoding/src/data.rs [hunk 1]
```diff
@@ -893,6 +953,17 @@ impl DataBlock {
...
+    as_type_mut_ref!(as_nullable_mut_ref, Nullable, NullableDataBlock);
+    as_type_mut_ref!(as_fixed_width_mut_ref, FixedWidth, FixedWidthDataBlock);
+    as_type_mut_ref!(
+        as_fixed_size_list_mut_ref,
+        FixedSizeList,
+        FixedSizeListBlock
+    );
+    as_type_mut_ref!(as_variable_width_mut_ref, VariableWidth, VariableWidthBlock);
+    as_type_mut_ref!(as_struct_mut_ref, Struct, StructDataBlock);
+    as_type_mut_ref!(as_dictionary_mut_ref, Dictionary, DictionaryDataBlock);
```
Comment (westonpace) [inline rust/lance-encoding/src/data.rs:966]: Can we call the function / macro name `..._mut` and not `..._mut_ref` for consistency with things like `[Option](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_mut)`

File: rust/lance-encoding/src/data.rs [hunk 2]
```diff
@@ -290,6 +290,48 @@ impl FixedWidthDataBlock {
...
+        assert!(block.bits_per_offset == 32);
+
+        let offsets = block.offsets.borrow_to_typed_slice::<u32>();
+        let offsets = offsets.as_ref();
+
+        let start_offset = offsets[selection.start as usize];
+        let end_offset = offsets[selection.end as usize];
+        self.bytes
+            .extend_from_slice(&block.data[start_offset as usize..end_offset as usize]);
+        self.offsets.push(self.bytes.len() as u32);
```
Comment (westonpace) [inline rust/lance-encoding/src/data.rs:319]: I'm not sure this is correct.  Shouldn't you be extending from `block.offsets` and not pushing a single value?

File: rust/lance-encoding/src/data.rs [hunk 3]
```diff
@@ -290,6 +290,48 @@ impl FixedWidthDataBlock {
     }
 }
 
+pub struct VariableWidthDataBlockBuilder1 {
```
Comment (westonpace) [inline rust/lance-encoding/src/data.rs:293]:
```suggestion
pub struct VariableWidthDataBlockBuilder {
```

**Review by westonpace — CHANGES_REQUESTED** [review]
Can you add some test cases?  Even just a basic simple case would be good.

**Comment by broccoliSpicy** [comment]
thanks for the review, I merged the code in this PR in PR #3099 and will close this one.

**Pull request closed without merging**